xbmc/addons/kodi-dev-kit/include/kodi/AddonBase.h

```diff
@@ -139,39 +139,75 @@ class ATTRIBUTE_HIDDEN CSettingValue
   ///@{
 
   /// @brief To get settings value as string.
-  std::string GetString() const { return (const char*)m_settingValue; }
+  std::string GetString() const { return str; }
 
   /// @brief To get settings value as integer.
-  int GetInt() const { return *(const int*)m_settingValue; }
+  int GetInt() const { return atoi(str.c_str()); }
 
   /// @brief To get settings value as unsigned integer.
-  unsigned int GetUInt() const { return *(const unsigned int*)m_settingValue; }
+  unsigned int GetUInt() const { return atoi(str.c_str()); }
 
   /// @brief To get settings value as boolean.
-  bool GetBoolean() const { return *(const bool*)m_settingValue; }
+  bool GetBoolean() const { return atoi(str.c_str()) > 0; }
 
   /// @brief To get settings value as floating point.
-  float GetFloat() const { return *(const float*)m_settingValue; }
+  float GetFloat() const { return static_cast<float>(atof(str.c_str())); }
 
   /// @brief To get settings value as enum.
   /// @note Inside settings.xml them stored as integer.
   template<typename enumType>
   enumType GetEnum() const
   {
-    return static_cast<enumType>(*(const int*)m_settingValue);
+    return static_cast<enumType>(GetInt());
   }
 
   ///@}
 
 private:
-  const void* m_settingValue;
+  const std::string str;
 };
 ///@}
 //------------------------------------------------------------------------------
 
 namespace addon
 {
 
+struct ATTR_DLL_LOCAL CPrivateBase
+{
+  /* Global variables of class */
+  static AddonGlobalInterface*
+      m_interface; // Interface function table to hold addresses on add-on and from kodi
+};
+
+class ATTR_DLL_LOCAL IInstanceInfo
+{
+public:
+  explicit IInstanceInfo(KODI_ADDON_INSTANCE_STRUCT* instance) : m_instance(instance) {}
+
+  KODI_ADDON_INSTANCE_TYPE GetType() const { return m_instance->info->type; }
+
+  bool IsType(KODI_ADDON_INSTANCE_TYPE type) const { return m_instance->info->type == type; }
+
+  uint32_t GetNumber() const { return m_instance->info->number; }
+
+  std::string GetID() const { return m_instance->info->id; }
+
+  std::string GetAPIVersion() const { return m_instance->info->version; }
+
+  bool FirstInstance() const { return m_instance->info->first_instance; }
+
+  operator KODI_ADDON_INSTANCE_STRUCT*() { return m_instance; }
+
+  operator KODI_ADDON_INSTANCE_STRUCT*() const { return m_instance; }
+
+private:
+  IInstanceInfo() = delete;
+  IInstanceInfo(IInstanceInfo&) = delete;
+  IInstanceInfo(const IInstanceInfo&) = delete;
+
+  KODI_ADDON_INSTANCE_STRUCT* m_instance;
```

Comment:
would it make sense to switch tostd::unique_ptrhere. I have seen so many manual deletes in so many files....